base/sparse/sparsematrix.jl

```diff
@@ -1884,48 +1884,61 @@ function setindex!{Tv,Ti,T<:Real}(A::SparseMatrixCSC{Tv,Ti}, x, I::AbstractVecto
 end
 
 
-
 # Sparse concatenation
 
 function vcat(X::SparseMatrixCSC...)
+    Tv = promote_type(map(x->eltype(x.nzval), X)...)
+    Ti = promote_type(map(x->eltype(x.rowval), X)...)
+
+    vcat(map(x->convert(SparseMatrixCSC{Tv,Ti}, x), X)...)
```

Comment:
I shouldn't have said "convert twice," I should have said "copy twice."

To be clearer, my suggestion is:
```julia
function vcat(X::SparseMatrixCSC...)
    Tv = promote_type(map(x->eltype(x.nzval), X)...)
    Ti = promote_type(map(x->eltype(x.rowval), X)...)
    <compute sizes and allocate outputs here>
    moffset = 0
    for x in X
        stuff!(colptr, rowval, nzval, moffset, x)
        moffset += size(x,1)
    end
    < pack as sparse matrix >
end
```

That way, even though `x` might be type-unstable, you don't really care, because it will dispatch to a type-specialized implementation of `stuff!`, and it's that function that has all the compute-intensive operations.